### DISCUSSION TITLE
Each user should sign his/her action with his/her own key

### DISCUSSION THREAD

**Discussion by Stezido** [openkfw/TruBudget#858] [Ideas]
## Summary 💡
Currently each organization has a walletaddress with private key which is used to sign every transaction the organization made.
It should be possible to sign a transaction with the user's walletaddress/privkey.

Maybe a network can switch between "Sign with organization wallet address" and "Sign with user wallet address" ?

## TODOs
- Create a Trubudget network that signs transactions with user's wallet address
- Create master-node which defines which signing must be used in a network (Maybe not possible?)

**Comment by georgimld** [comment]
This would probably be a breaking change but it should be possible to create a chain with different parameters and users could decide if they want to use the user's wallet address or the node's wallet address

**Comment by georgimld** [comment]
Technically this is possible by following the guidelines in Multichain's [External key Management Tutorial](https://www.multichain.com/developers/external-key-management/). 

However this is a little more tricky to implement in TruBudget. 

Possible Solutions:

1. **Good security, bad usability:** Each user has an address, imported in the node wallet when they are created. The private key is encrypted on the chain using the user's password (which is only known to the user). This way, no one has access to the private key except the user. However, when performing any writing operation on the chain, the user has to be prompted for their password, in order to decrypt the private key and sign the transaction.

1. **Not so good security, better usability**: The address and the private key of each user is imported in the wallet, so the user doesn't need to be prompted for their password every time a writing operation is performed. However everyone who has access to the node has access to all private keys, so the security is not optimal.

**Reply by georgimld** [reply]
Solution 2 is implemented in #978

Solution 1 can be implemented at a later time if a higher level of security is required in TruBudget
